### PR TITLE
update relationship routes and controllers

### PR DESCRIPTION
**Pull request by ally0426**
nested routes

### CODE REVIEW

**Comment by ally0426** [comment]
Testing works (e.g. http://localhost:3000/users/3/apps/2)

File: backend/app/controllers/users_controller.rb [hunk 1]
```diff
@@ -16,18 +16,33 @@ def show
   # POST /users
   def create
     @user = User.create(user_params)
+    if(@user.save)
+      redirect_to @user
+    else
+      render 'new'
+    end
```
Comment (marmalade82) [inline backend/app/controllers/users_controller.rb:23]: This code would typically work if the frontend was also being served by Rails, but that isn't the case with our app. This block isn't needed -- you should just send the JSON response, and the Gatsby frontend will decide where to redirect to.

File: backend/app/controllers/users_controller.rb [hunk 2]
```diff
@@ -16,18 +16,33 @@ def show
...
     json_response(@user, :created)
   end
 
+  def edit
+  
+  end
+
```
Comment (marmalade82) [inline backend/app/controllers/users_controller.rb:30]: We don't need an edit endpoint. The `edit` endpoint is for Rails to send the HTML form for editing. But in our app, Gatsby is serving the form.

File: backend/app/controllers/users_controller.rb [hunk 3]
```diff
@@ -16,18 +16,33 @@ def show
...
+            render 'edit'
+        end
+    #@user.update!(user_params)
     head :no_content
   end
 
   # DELETE /users/:id
   def destroy
     @user.destroy
+    redirect_to users_path
```
Comment (marmalade82) [inline backend/app/controllers/users_controller.rb:45]: The rails backend is not responsible for rendering or redirecting (except for authentication). The Gatsby frontend code is responsible, which is why these changes aren't needed.

File: backend/app/models/user.rb [hunk 4]
```diff
@@ -1,5 +1,5 @@
 class User < ApplicationRecord
-  has_many :apps
+  has_many :apps #, tags
```
Comment (marmalade82) [inline backend/app/models/user.rb:2]: Why is this commented out? Users do have tags.

File: backend/app/controllers/apps_controller.rb [hunk 5]
```diff
@@ -2,32 +2,54 @@ class AppsController < ApplicationController
...
   def update
-    @app.update!(app_params)
+    @user = User.find(params[:user_id])
+    @app = @user.apps.find(params[:id])
+    if(@app.update(app_params))
+      redirect_to @app
+    else
+      render 'edit'
+    end
+    #@app.update!(app_params)
```
Comment (marmalade82) [inline backend/app/controllers/apps_controller.rb:42]: The controller shouldn't be redirecting or rendering any pages, because that is Gatsby's responsibility. For example, `redirect to @app` and `render 'edit'` don't belong in this file.

File: backend/app/controllers/apps_controller.rb [hunk 6]
```diff
@@ -2,32 +2,54 @@ class AppsController < ApplicationController
   before_action :set_app, only: [:show, :update, :destroy]
 
 
-  # GET /apps
+  # GET /users/:user_id/apps
   def index
-    @apps = App.all
+    @user = User.find(params[:user_id])
+    @apps = @user.apps.all
+    #@apps = App.all
```
Comment (marmalade82) [inline backend/app/controllers/apps_controller.rb:9]: Don't comment out the line. Just delete it.

File: backend/config/routes.rb [hunk 7]
```diff
@@ -13,5 +13,11 @@
...
 
   # CRUD for Apps, Users, & Tags
-  resources :apps, :users, :tags
+  resources :users do
+    resources :apps 
+  end
+
+  resources :apps do
+    resources :tags
+  end
```
Comment (marmalade82) [inline backend/config/routes.rb:22]: I understand what you're going for here, and it makes sense. But if you do it this way, it creates too many conflicting resources.

For example, both `users/:user_id/apps` and `/apps` are both endpoints that use the same controller action. If you run `rails routes` in your console, you'll see the problem I've posted below: both endpoints use the `apps#index` action, which is a bad conflict.

This kind of conflicts need to be fixed. Take a look at https://guides.rubyonrails.org/routing.html#nested-resources section 2.7.2 about shallow nesting, which might be needed to fix this.

controller_conflicts

**Comment by ally0426** [comment]
Yeah, need to restore localhost:3000/apps for anyone who visits the site.

**Comment by marmalade82** [comment]
@ally0426 Okay. Maybe we can figure this out during one of the standups

**Comment by marmalade82** [comment]
@ally0426 Closing, as most of these changes may no longer be needed. If you feel there's a strong need for them, we can go over them in the Sprint Planning meeting.
